gh-60191: Implement ast.compare

isidentical opened the pull request:
<!-- gh-issue-number: gh-60191 -->
* Issue: gh-60191
<!-- /gh-issue-number -->

isidentical committed:
bpo-15987: Implement ast.compare

hansrajdas commented on Lib/ast.py:
```diff
@@ -368,6 +368,69 @@ def walk(node):
+):
+    """
+    Recursively compare given 2 AST nodes. If *compare_types* is False, the
+    field values don't be checked whater they belong to same type or not. If
```
Typo - I think "whater" should be "whether". In above doc file also

isidentical replied:
Yes, thanks you.

hansrajdas commented on Lib/ast.py:
```diff
@@ -368,6 +368,69 @@ def walk(node):
+            for a_item, b_item in zip(a, b):
+                if not _compare(a_item, b_item):
+                    return False
+            else:
```
We can remove this else and move return statement parallel to for...
At other places also

isidentical replied:
I disagree, the whole purpose of `else` is this.

hansrajdas commented on Lib/ast.py:
```diff
@@ -368,6 +368,69 @@ def walk(node):
+    if type(a) is not type(b):
+        return False
+    if compare_fields:
+        if not _compare_member("_fields"):
```
Combining both if conditions with `and` boolean operator won't make too complex...
```python
    if compare_fields and not _compare_member("_fields"):
        return False
```
Below also

isidentical committed:
unwrap ifs a level

iritkatriel commented on Lib/ast.py:
```diff
@@ -368,6 +368,67 @@ def walk(node):
+            )
+        elif isinstance(a, list):
+            if len(a) != len(b):
+                return False
```
I think you're assuming here that b is also a list, but if compare_types is False is might not be?

isidentical replied:
In theory, if you didn't alter the AST by your hand, a sequence field is always a list. So when comparing 2 different ASTs, if the field 'X' of AST 'A' has the type list, then the field 'X' of AST 'B' should be also a list (this is only for sequence types thought, not for optionals).

jeremyhylton replied:
compare_types is about the types of Python objects, while the list here is about the Python list used for fields that have repeated values.

AlexWaygood committed:
Merge branch 'main' into bpo-15987

jeremyhylton commented on Doc/library/ast.rst:
```diff
@@ -2472,6 +2472,17 @@ effects on the compilation of a program:
+    field values won't be checked whether they belong to same type or not. If
+    *compare_fields* is ``True``, members of ``_fields`` attribute on both node's
+    type will be checked. If *compare_attributes* is ``True``, members of
+    ``_attributes`` attribute on both node's will be compared.
```
I think this needs to be spelled out a little more directly, with more words. "If *compare_types* is ``False``, the
    field values won't be checked whether they belong to same type or not." I think that means when the field values are compared, their values will be compared without regard to their type. But it literally seems to say that "field values won't be checked" which isn't right.

jeremyhylton commented on Lib/ast.py:
```diff
@@ -401,6 +401,67 @@ def walk(node):
+    def _compare_member(member):
+        for field in getattr(a, member):
+            if not hasattr(a, field) and not hasattr(b, field):
+                continue
```
I don't understand why it would be true that getattr(a, member) == field and not hasattr(a, field). (but I would be happy to understand why it could be true.)

jeremyhylton committed:
A few revision to clarify some subtleties of comparing AST objects.

Use separate helper functions to compare _fields and _attributes, because
_attributes are always simple strings.

Move the type comparison test to the point where it is relevant,
rather than making unnecessary type tests.

Add comments to explain the logic in more detail.

jeremyhylton reviewed (commented):
I added a few comments, but realized I wanted to try a slightly revised implementation to answer the questions I had. I'll share the revision in a moment.

jeremyhylton committed:
Merge remote-tracking branch 'upstream/main' into pr_19211

jeremyhylton committed:
Remove the compare_fields option.

The basic functionality here is to compare asts recursively. If
compare_fields is False, then the comparison is not recursive. It just
compares that the top-level AST objects are the same and have the same
attributes. This option doesn't seem interesting enough to offer as a
feature.

Revise the docstring for compare() to explain the options in more detail.

jeremyhylton committed:
Revise docstring and documentation for consistency.

jeremyhylton committed:
This PR now describes a feature of 3.14. Revise what's new / news docs.

iritkatriel commented on Doc/library/ast.rst:
```diff
@@ -2472,6 +2472,25 @@ effects on the compilation of a program:
+   attributes are ignored. Otherwise they must all be equal. This
+   option is useful to look for asts that are structurally equal but
+   might differ in whitespace or similar details.
+   .. versionadded:: 3.14
```
```suggestion

   .. versionadded:: 3.14
```

iritkatriel commented on Doc/library/ast.rst:
```diff
@@ -2472,6 +2472,25 @@ effects on the compilation of a program:
 
+.. function:: compare(a, b, /, *, compare_types=True, compare_attributes=False)
+
+   Recursively compares two ast nodes.
```
```suggestion
   Recursively compares two asts.
```

iritkatriel commented on Doc/whatsnew/3.14.rst:
```diff
@@ -86,6 +86,13 @@ New Modules
+ast
+---
+
+Added :func:`ast.compare` for comparing 2 AST node.
```
```suggestion
Added :func:`ast.compare` for comparing 2 ASTs.
```

iritkatriel commented on Doc/whatsnew/3.14.rst:
```diff
@@ -86,6 +86,13 @@ New Modules
+---
+
+Added :func:`ast.compare` for comparing 2 AST node.
+(Contributed by Batuhan Taskaya in :issue:`15987`)
```
```suggestion
(Contributed by Batuhan Taskaya and Jeremy Hylton in :issue:`15987`)
```

iritkatriel commented on Doc/whatsnew/3.9.rst:
```diff
@@ -1,3 +1,4 @@
+
```
This file doesn't need to change.

iritkatriel commented on Lib/ast.py:
```diff
@@ -401,6 +401,90 @@ def walk(node):
+    compare_types=True,
+    compare_attributes=False,
+):
+    """Recursively compares two ast nodes.
```
```suggestion
    """Recursively compares two ASTs.
```

jeremyhylton committed:
Update 3.9.rst

iritkatriel commented on Doc/library/ast.rst:
```diff
@@ -2472,6 +2472,25 @@ effects on the compilation of a program:
+   *compare_types* is ``True`` (default), then Constant objects must
+   have the same type and value to be equal. If *compare_types* is
+   ``False``, then Constant objects must only have the same values,
+   e.g. Constant(1.0) equals Constant(1).
```
As discussed, let's not do this.

jeremyhylton committed:
Remove the compare_types option, which seems unnecessary.

In practice the primary effect of this change was for Constant() nodes
where the value of two constants were equal without being the same
literal type, e.g. ast.compare(Constant(1), Constant(True),
compare_types=True) was True.

This behavior doesn't seem like it has a very strong motivation, and
adds some complexity. We can add it back later if there's a clear need
for it.

jeremyhylton committed:
Update Doc/whatsnew/3.14.rst

Co-authored-by: Irit Katriel <1055913+iritkatriel@users.noreply.github.com>

jeremyhylton committed:
Update Doc/whatsnew/3.14.rst

Co-authored-by: Irit Katriel <1055913+iritkatriel@users.noreply.github.com>

jeremyhylton committed:
Remove compare_types from doc. Change ast node to AST.

jeremyhylton committed:
Change AST node to AST.

jeremyhylton committed:
Merge remote-tracking branch 'upstream/main' into pr_19211

jeremyhylton committed:
Merge branch 'bpo-15987' into pr_19211

iritkatriel commented on Lib/ast.py:
```diff
@@ -401,6 +401,76 @@ def walk(node):
+    compare_attributes affects whether AST attributes are considered
+    in the comparison. If compare_attributes is False (default), then
+    attributes are ignored. Otherwise they must all be equal. This
+    option is useful to look for asts that are structurally equal but
```
```suggestion
    option is useful to check whether the ASTs are structurally equal but
```

iritkatriel commented on Doc/library/ast.rst:
```diff
@@ -2472,6 +2472,19 @@ effects on the compilation of a program:
+   *compare_attributes* affects whether AST attributes are considered
+   in the comparison. If compare_attributes is ``False`` (default), then
+   attributes are ignored. Otherwise they must all be equal. This
+   option is useful to look for asts that are structurally equal but
```
```suggestion
   option is useful to check whether the ASTs are structurally equal but
```

iritkatriel commented on Lib/ast.py:
```diff
@@ -401,6 +401,76 @@ def walk(node):
+        else:
+            return type(a) is type(b) and a == b
+
+    def _compare_fields():
```
Should there be a check that all fields of b are also fields of a?

iritkatriel replied:
```suggestion
    def _compare_fields(a, b):
```

iritkatriel commented on Lib/ast.py:
```diff
@@ -401,6 +401,76 @@ def walk(node):
+
+    def _compare_attributes():
+        # Attributes are always strings.
+        for attr in a._attributes:
```
As above - check that a has all the attributes of b?

jeremyhylton replied:
Per the check on line 463, we know that a and b are the same type. Thus they are required to have the same fields and attributes. (Apparently this wasn't always true but was probably a change Jelle introduced in 3.13.

jeremyhylton replied:
but user could modify _attributes per our f2f conversaton

jeremyhylton committed:
One more change of "ast nodes" to "ASTs"

iritkatriel commented on Lib/ast.py:
```diff
@@ -401,6 +401,76 @@ def walk(node):
+    if type(a) is not type(b):
+        return False
+    # a and b are guaranteed to have the same type, so they must also
+    # have identical values for _fields and _attributes.
```
I think since _fields etc are mutable we might not want to assume that the ASTs are well formed and have the correct _fields and _attributes.

iritkatriel commented on Lib/ast.py:
```diff
@@ -401,6 +401,76 @@ def walk(node):
+        else:
+            return True
+
+    def _compare_attributes():
```
```suggestion
    def _compare_attributes(a, b):
```

iritkatriel commented on Lib/ast.py:
```diff
@@ -401,6 +401,76 @@ def walk(node):
+        return False
+    # a and b are guaranteed to have the same type, so they must also
+    # have identical values for _fields and _attributes.
+    if not _compare_fields():
```
```suggestion
    if not _compare_fields(a, b):
```

iritkatriel commented on Lib/ast.py:
```diff
@@ -401,6 +401,76 @@ def walk(node):
+    # have identical values for _fields and _attributes.
+    if not _compare_fields():
+        return False
+    if compare_attributes and not _compare_attributes():
```
```suggestion
    if compare_attributes and not _compare_attributes(a, b):
```

jeremyhylton committed:
Merge AST comparison into the test for AST validation.

The chief benefit is that we only have to parse a large set of modules
once. We can test parse and compare at the same time.

jeremyhylton committed:
Merge tests for AST parsing and comparison.

This change means that we only need to parse the full set of modules
from the standard library once.

iritkatriel commented on Lib/test/test_ast.py:
```diff
@@ -1066,6 +1071,100 @@ def test_ast_asdl_signature(self):
+        if support.is_resource_enabled("cpu"):
+            files = STDLIB_FILES
+        else:
+            files = random.sample(STDLIB_FILES, 10)
```
Will this make debugging test failures harder because the test is not deterministic?

jeremyhylton replied:
Yeah, I didn't like that either. I deleted this whole section.

iritkatriel commented on Lib/test/test_ast.py:
```diff
@@ -1066,6 +1071,100 @@ def test_ast_asdl_signature(self):
+                    ast.compare(a, b), f"{ast.dump(a)} != {ast.dump(b)}"
+                )
+
+    def test_compare_tests(self):
```
```suggestion
    def test_compare_modes(self):
```

iritkatriel commented on Lib/test/test_ast.py:
```diff
@@ -1066,6 +1071,100 @@ def test_ast_asdl_signature(self):
+                    ast.compare(a, b), f"{ast.dump(a)} != {ast.dump(b)}"
+                )
+
+    def test_compare_options(self):
```
```suggestion
    def test_compare_attributes_option(self):
```

iritkatriel commented on Doc/whatsnew/3.14.rst:
```diff
@@ -86,6 +86,13 @@ New Modules
+ast
+---
+
+Added :func:`ast.compare` for comparing 2 ASTs.
```
```suggestion
Added :func:`ast.compare` for comparing two ASTs.
```

jeremyhylton committed:
Merge branch 'main' into bpo-15987

jeremyhylton committed:
Improve description of compare_attributes arg

Co-authored-by: Irit Katriel <1055913+iritkatriel@users.noreply.github.com>

jeremyhylton committed:
AP style: Spell out numbers under 10

Co-authored-by: Irit Katriel <1055913+iritkatriel@users.noreply.github.com>

jeremyhylton committed:
Improve description of compare_attributes arg

Co-authored-by: Irit Katriel <1055913+iritkatriel@users.noreply.github.com>

jeremyhylton committed:
Improve robustness of compare() in the face of user-modification of AST.

The comparison fundamentally depends on _fields and _attributes, which
could be modified by the user. It's not clear that such modifications
are sensible or supported by the API, but we can at least make sure
comparison doesn't silently ignore those comparisons.

Also pass a and b as arguments to helper methods instead of using them
from the enclosing scope.

jeremyhylton committed:
Update Lib/test/test_ast.py

Co-authored-by: Irit Katriel <1055913+iritkatriel@users.noreply.github.com>

jeremyhylton committed:
Update Lib/test/test_ast.py

Co-authored-by: Irit Katriel <1055913+iritkatriel@users.noreply.github.com>

iritkatriel commented on Lib/test/test_ast.py:
```diff
@@ -1066,6 +1069,113 @@ def test_ast_asdl_signature(self):
+        self.assertFalse(ast.compare(b, a))
+
+        b._fields = a._fields
+        b.spam = "Spam"
```
```suggestion
        b.spam = a.spam
```

(for consistency with the way it's done below)

iritkatriel commented on Lib/test/test_ast.py:
```diff
@@ -1066,6 +1069,113 @@ def test_ast_asdl_signature(self):
+            return ast.parse(a), ast.parse(b)
+
+        a, b = parse("2 + 2", "2+2")
+        self.assertTrue(ast.compare(a, b, compare_attributes=False))
```
```suggestion
        self.assertTrue(ast.compare(a, b))
        self.assertTrue(ast.compare(a, b, compare_attributes=False))
```

jeremyhylton committed:
Update Lib/test/test_ast.py

Co-authored-by: Irit Katriel <1055913+iritkatriel@users.noreply.github.com>

jeremyhylton committed:
Update Lib/test/test_ast.py

Co-authored-by: Irit Katriel <1055913+iritkatriel@users.noreply.github.com>

iritkatriel commented on Misc/NEWS.d/next/Library/2020-03-28-21-00-54.bpo-15987.aBL8XS.rst:
```diff
@@ -0,0 +1,3 @@
+Implemented :func:`ast.compare` for comparing two ASTs. Patch by Batuhan
+Taskaya with some help from Jeremy Hylton.
+
```
```suggestion
```

iritkatriel committed:
whitespace

jeremyhylton committed:
Merge branch 'main' into bpo-15987

jeremyhylton committed:
Merge branch 'main' into bpo-15987

JelleZijlstra commented on Doc/library/ast.rst:
```diff
@@ -2472,6 +2472,19 @@ effects on the compilation of a program:
+   Recursively compares two ASTs.
+
+   *compare_attributes* affects whether AST attributes are considered
+   in the comparison. If compare_attributes is ``False`` (default), then
```
```suggestion
   in the comparison. If *compare_attributes* is ``False`` (default), then
```

JelleZijlstra commented on Doc/library/ast.rst:
```diff
@@ -2472,6 +2472,19 @@ effects on the compilation of a program:
+   in the comparison. If compare_attributes is ``False`` (default), then
+   attributes are ignored. Otherwise they must all be equal. This
+   option is useful to check whether the ASTs are structurally equal but
+   might differ in whitespace or similar details.
```
We could be more explicit in saying that "attributes" includes the line and column numbers.

JelleZijlstra commented on Lib/ast.py:
```diff
@@ -401,6 +401,78 @@ def walk(node):
+                a,
+                b,
+                compare_attributes=compare_attributes,
+
```
```suggestion
```
Extra blank line

JelleZijlstra commented on Lib/ast.py:
```diff
@@ -401,6 +401,78 @@ def walk(node):
+    def _compare_attributes(a, b):
+        if a._attributes != b._attributes:
+            return False
+        # Attributes are always strings.
```
Aren't they always ints?

jeremyhylton committed:
Update Doc/library/ast.rst

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

jeremyhylton committed:
Update Lib/ast.py

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

jeremyhylton committed:
Attributes are ints not strings.

jeremyhylton committed:
Explain examples of what are included in attributes.

JelleZijlstra commented on Doc/library/ast.rst:
```diff
@@ -2472,6 +2472,20 @@ effects on the compilation of a program:
+   in the comparison. If *compare_attributes* is ``False`` (default), then
+   attributes are ignored. Otherwise they must all be equal. This
+   option is useful to check whether the ASTs are structurally equal but
+   differ in whitespace or similar details. Attributes include numbers
```
```suggestion
   differ in whitespace or similar details. Attributes include line numbers
```

jeremyhylton committed:
Update Doc/library/ast.rst

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

jeremyhylton committed:
Merge branch 'main' into bpo-15987

jeremyhylton committed:
Merge branch 'main' into bpo-15987
